src/devtools/client/debugger/src/reducers/source-actors.js

```diff
@@ -101,6 +105,31 @@ function updateBreakableLines(state, action) {
   return updateResources(state, [{ id: sourceId, breakableLines: value }]);
 }
 
+function updateBreakpointHitCounts(state, action) {
+  const value = asyncActionAsValue(action);
+  if (value.state === "pending") {
+    return state;
+  }
+
+  const hitCountsMap = value.value.hits.reduce((acc, { location, hits }) => {
+    acc[location.line] = { ...(location.line || {}), [location.column]: hits };
```

Comment:
:hand: Strictly speaking I don't think we have to make a copy here. We're starting with a fresh `acc` object and providing new empty objects inside the loop. That means it _is_ safe to literally mutate those line objects inside of this reducer, because they're brand new.  Not sure on the _actual_ perf cost here, but I think this would technically be a bit faster:

```ts
if (!acc[location.line]) {
  acc[location.line] = {}
}

acc[location.line][location.column] = hits;
return acc;
```

(yes I hate the "fill in a default value" logic here - really wish JS had an equivalent of Python's `defaultdict` type)

Reply:
yeah, your aversion to the `if` is exactly why I did it this way, but you're right, we should try not to make thousands of copies for no reason!